### PR TITLE
[yaireo__tagify] Fix types, add missing types, add missing docs, cleanup

### PR DESCRIPTION
**Pull request by blutorange**
Woah, the devil is in the details once you actually start to compare the docs, the source code and how the program behaves in practice. The hardest part is deciding which differences / omissions in the docs are intentional... I hope I didn't make any mistakes, but if anyone could review, that would be great.

* Extracted nested objects into their own interfaces (e.g. `Tagify.TagifySettings.autocomplete` > `Tagify.AutocompleteSettings`). Personally I find it very helpful to have these types available for refactoring your code, e.g. when you want to create a `function createAutocompleteSettings` or `const autocompleteSettings = ...`
* Refactored the events and their data into an EventDataMap interface. This also allows us to reuse that event data map for typing the `off` method and the `callbacks` setting.
    * I also added the events `change` and `dropdown:noMatch`, see https://github.com/yairEO/tagify#events
* Also, I edited the docs a little, correcting some typos, adding docs where it was missing, and added `@param` / `@returns` tags.
* Added an optional type parameter `T` to the Tagify class that encodes the type of the tag data. It must extend `BaseTagData`, which is `{value: string}` and defaults to `TagData`, which is `{value: string, [key: string]: any}`. This ensures compatibility but allows people to be more strict if they wish.
    * This also allows us to type the parameter of `addEmptyTag` as `Partial<T>`
* Adjusted the types of a few settings and added missing settings:
    * Added the `hooks` option to `TagifySettings.hooks`, see https://github.com/yairEO/tagify#hooks
    * `pattern` can be `null`
    * `dropdown.appendTarget` can be `null` (which results in the dropdown being appended to `document.body`)
    * `dropdown.position` defaults to `all`, not `null`
    * `dropdown.highlightFirst` defaults to `false`, not `null`
    * Added the `dropdown.placeAbove` option to the `TagifySettings`, see https://github.com/yairEO/tagify#settings
    * Added the `data: {value: string}` parameter to `dropdownItemNoMatch` callback, see e.g. https://yaireo.github.io/tagify/#section-advance-options and https://github.com/yairEO/tagify/blob/master/src/parts/dropdown.js#L44
    * `mixTagsInterpolator` is not a string array, but a tuple of two elements,  so I changed the type to `[string, string]`
* Adjusted the types of a few methods:
    * `Tagify.addTags` always seems returns an array of added tags, even when passed a single string. I tried it out and this also seems to match the source code https://github.com/yairEO/tagify/blob/master/src/tagify.js#L1196
    * First parameter of `Tagify.getWhitelistItem` must be a string, not `TagData` (despite the documentation indicating otherwise), see https://github.com/yairEO/tagify/blob/master/src/tagify.js#L814
    * Added parameters `property` and `whitelist` to `Tagify.getWhitelistItem`
    * `Tagify.getTagElmByValue` returns `undefined` when it does not find a tag.
    * `Tagify.isTagDuplicate` does not alway return a number, but sometimes also `false`, see https://github.com/yairEO/tagify/blob/master/src/tagify.js#L752
    * Split `Tagify.tagData` into two method signatures, one for the getter and one for the setter.
    * For the setter, the second argument is a `Partial<TagData>` -- it assigns the given properties and keeps the properties that are not given.
    * Added a third method signature for the third optional parameters `override` (completely replace the tag data).
    * Tagify.editTag` and `Tagify.injectAtCaret` return `this`, not `void`
    * Second argument `force` of `Tagify.toggleClass` is optional.
    * Refactored the `parseTemplate` type signature so that (a) it just reads the available methods from `Tagify.TagifySettings` and (b) uses a type parameter for the arguments of the custom template function to get rid of `any`.
* Replaced the tests for the `templates` callback methods with the defaults from https://github.com/yairEO/tagify/blob/master/src/parts/templates.js. These should pass the type check, but I found a few missing types
   * These `templates` callback methods are invoked with the this context set to the tagify instance. Since the this context is used even in the defaults, I added `this: Tagify` to the callback type signatures.
   * There's apparently a property `TagifySettings#classNames`, which I added. See https://github.com/yairEO/tagify#drag--sort and https://github.com/yairEO/tagify/blob/master/src/parts/defaults.js#L33
   * `required` and `readonly` can be set via HTML attributes, but are also available in the `TagifySettings`. It's a bit hidden, but can be found here in the docs: https://github.com/yairEO/tagify#html-input--textarea-attributes However, even though the docs suggest you can set these options by passing them to the constructor, that is not actually the case if you try it out / read the code. These two options are added to the `TagifySettings` when the tagify instance is created. So I split the settings into `TagifySettings` and `TagifyConstructorSettings`. The latter are used in the constructor, the former everywhere else.
* This is missing in the docs, but when we actually try to use the `tagify.dropdown.*` methods (such as `selectAll`), we notice they won't work. This is because they have to be called with the tagify instance as the this context. Just calling `tagify.dropdown.selectAll()` will result in an error. I added the this context in the method signature and adjusted the tests.
    * I also added the `refilter` method to the dropdown methods.

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/yairEO/tagify and https://yaireo.github.io/tagify
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Review by Brakebein — CHANGES_REQUESTED** [review]
Wow, this is almost a complete rewrite of the type declarations. It looks like you really reverse engineered every method. I agree it's a cumbersome task, since code and docs are not always in sync. For the initial declarations I didn't explicitly test every method, I just looked at the code  (and docs). So, some declaration might not be 100 % correct.

And I'm happy to welcome you as a additional owner. Because the problem with being the only owner is that you have to wait for a DT maintainer to approve any changes, which can take a while. As far as I understood the DT bot and read elsewhere, you are not able to approve your own changes. Now, being two owners, PRs should be approved and merged much faster.

I made a PR a week ago (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51447), where I addressed `dropdown.position` and `dropdown.placeAbove` settings. But for the just mentioned reason, this PR is still open. I will close it in favor of your PR.

Your changes seem very solid and well done.
There is quite a big collection of interfaces now. I'm only not sure about the `Mode` type and `Dom` interface: these are very generic names that may cause name conflicts when importing into your code. Of course, you could import with `as` alias, but I try to avoid this (maybe a matter of personal taste). And are they really needed? It's only my opinion. If you think it's okay this way, I will approve it.

A few request changes:

* `removeAllTags()` has an optional argument `{ withoutChangeEvent: boolean }`. It's only used internally, but users calling this method might also want to set this flag. See my PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51447 and https://github.com/yairEO/tagify/blob/master/src/tagify.js#L1450-L1465
* Can you also add `update()` and `getMixedTagsAsString()`, please? I was recently using these methods. https://github.com/yairEO/tagify/blob/master/src/tagify.js#L1480-L1539

**Comment by blutorange** [comment]
@Brakebein Thanks for your review. It might seem like a rewrite, but it's not as much as you might think. Your initial declarations are still mostly there, just a little bit refactored. That was certainly easier that starting from scratch : ) And yes, that's why I added myself to the owners so that there are a few more people who can review and approve changes.

I did check the source code and the docs, though I did not include some optional parameters that did look like they were only meant to be used internally. But I think I might take another look at the source code and check again, we should also consider whether there's a use case for those parameters. Especially since mentioned that you've got a use case. If we document a method at all, it's probably better to document it completely.

Interfaces: I used rather generic names since it was in its own namespace, but you're right, they're imported directly without the namespace. So yes, I should definitely change the name to something more specific. Regarding whether they are necessary: I can inline them again if we decide it's better, but my reasons for doing so are:
    * Personally I find it rather helpful when packages provide these types, so I can use them for utility and helper methods etc.
    * For example, when I use tagify, I can create a helper function `createDropDownOptions`  and use the type as the return type.
    *  Just recently, I updated the types for `spectrum` and it had an option `appendTo`, which was passed to JQuery's appendTo function. That argument of that function is a union of many HTML-like types, it would have helped if JQuery had exposed a type like `ElementLike`.
    *  Extracting nested types into their own interfaces also helps with adding documentation to them.

**Comment by Brakebein** [comment]
@blutorange I'm fine with all the interfaces. I had only doubts about `Mode` and `Dom`. So, renaming them should be fine.

For `replaceAllTags()`, I don't have a personal use case. When I compared the last releases, I saw that this optional argurment has been added. Even it is not official documented in the readme or elsewhere, I think it is good, if a method is declared, to declare all possible parameters.

For `update()` and `getMixedTagsAsString()`, I thought they are only used internally, so didn't include them in the declarations. But now I have the use case where they are needed: Changes to the `originalInput.value` are only made when a tag is added (or edited or removed). In `mode: 'mix'` however, if you type some normal text, the value is not updated. So, you need to call `update()` and `getMixedTagsAsString()` manually to get the final string.

I just closed the other PR.

**Comment by blutorange** [comment]
Making the changes now, should be done soon. There's one issue with the `removeAllTags` method, however: it's supposed to take a `{withoutChangeEvent:boolean}`, but the implementation is broken: https://github.com/yairEO/tagify/blob/master/src/tagify.js#L1451-L1464

It's set to an object at the start of the function, and then passed to `update` at the end -- but wrapped in an additional object `{withoutChangeEvent:{withoutChangeEvent:boolean}}`. The effect is that the method behaves as if the options were always set to `true` and never actually triggers the change event, as you can see in the playground

image

We might consider opening an issue for this, but for now I won't include that option until it gets implemented properly. It's kinda a good example why TypeScript helps avoiding bugs though ; )

**Comment by blutorange** [comment]
* Renamed type
    * `Mode` to `TagifyMode`
    * `Dom` to `DomReference`
* Added method `Tagify.update(opts?: Tagify.UpdateOptions): void`
* Added method `Tagify.getMixedTagsAsString(): string`

**Comment by blutorange** [comment]
@Brakebein Would also be great if you could check this PR against your code you mentioned where you need the additional methods and see if there are no type errors (easiest way would be to temporarily replace the file in node_modules/@types/... with the file from this PR)

File: types/yaireo__tagify/index.d.ts
Comment (Brakebein) [inline types/yaireo__tagify/index.d.ts:1132]: Nice description. There is only one backtick to much.

**Review by Brakebein — CHANGES_REQUESTED** [review]
Thanks for adding these methods. There is only one backtick to much in the description.

I think, we can then approve this PR, as we don't know how long it takes until the issue with `removeAllTags()` will be resolved. We can update the type declarations later on. Or do you want to wait?

**Comment by blutorange** [comment]
@Brakebein  Thanks for pointing that out. Yeah, we can merge this now. If / when that issue is resolved, we can make another PR and update the version to 3.24. For now, these are the declarations for 3.23.

**Review by Brakebein — APPROVED** [review]
Thanks for your huge amount of work!

**Comment by Brakebein** [comment]
Ready to merge

**Pull request merged**
